File: my-submissions/1798-max-number-of-k-sum-pairs/solution.java
class Solution {
    public int maxOperations(int[] nums, int k) {
        int count = 0;
        Arrays.sort(nums);

        // System.out.println("nums => " + Arrays.toString(nums));

        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            int sum = nums[left] + nums[right];

            if (sum == k) { // valid pair
                left++;
                right--;
                count++;
            } else if (sum < k) {
                left++;
            } else {
                right--;
            }

        }

        return count;
    }
}

File: my-submissions/1798-max-number-of-k-sum-pairs/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxOperations = function (nums, k) {
    let count = 0;
    nums.sort((a, b) => a - b);

    let left = 0;
    let right = nums.length - 1;

    while (left < right) {
        let sum = nums[left] + nums[right];

        // found a valid pair
        if (sum === k) {
            left++;
            right--;
            count++;
        } else if (sum < k) {
            // current sum too small, move left pointer to larger value
            left++;
        } else {
            // current sum too large, move right pointer to smaller value
            right--;
        }
    }
    return count;
};
